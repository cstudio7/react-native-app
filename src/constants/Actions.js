export default {
  favoriteName: 'FAVORITE_NAME',
  unfavoriteName: 'UNFAVORITE_NAME',
  saveNames: 'SAVE_NAMES'
};