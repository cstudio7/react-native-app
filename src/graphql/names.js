import gql from 'graphql-tag';

export const femaleNamesQuery = gql`
  {
    femaleNames {
      id
      name
    }
  }
`;

export const maleNamesQuery = gql`
  {
    maleNames {
      id
      name
    }
  }
`;
