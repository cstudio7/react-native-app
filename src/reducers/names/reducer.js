const R = require('ramda');

import {
  LISTSCREEN_NAME_FAVORITE,
  FAVSCREEN_NAME_FAVORITE,
  LISTSCREEN_NAME_UNFAVORITE,
  FAVSCREEN_NAME_UNFAVORITE,
  FETCH_NAMES_SUCCESS
} from '../../constants/constants';
import createReducer from '../../modules/createReducer/createReducer';

const initialState = {
  female: [],
  male: []
};

const updateItemInArray = (state, name, gender) => {
  const genderData = state[gender];
  const index = R.findIndex(R.propEq('id', name.id))(genderData);

  return {
    ...state,
    [gender]: [
      ...genderData.slice(0, index),
      name,
      ...genderData.slice(index + 1)
    ]
  };
};

const mergeGenderNamesWithState = (state, response) => {
  const gender = response.femaleNames ? 'female' : 'male';
  if (!R.path([gender, 'length'], state)) {
    return {
      ...state,
      [gender]: response.femaleNames || response.maleNames
    };
  }

  return {
    ...state,
    [gender]: (response.femaleNames || response.maleNames).map(name =>
      R.mergeAll([
        name,
        state[gender].find(stateName => stateName.name === name.name)
      ])
    )
  };
};

const saveNames = (state, action) =>
  mergeGenderNamesWithState(state, action.response);

const favoriteName = (state, action) => {
  const { payload, type } = action;
  const newName = {
    ...payload.name,
    isFavorite:
      type === FAVSCREEN_NAME_FAVORITE || type === LISTSCREEN_NAME_FAVORITE
  };
  return updateItemInArray(state, newName, payload.gender);
};

const unfavoriteName = favoriteName;

export default createReducer(initialState, {
  [FETCH_NAMES_SUCCESS]: saveNames,
  [LISTSCREEN_NAME_FAVORITE]: favoriteName,
  [FAVSCREEN_NAME_FAVORITE]: favoriteName,
  [LISTSCREEN_NAME_UNFAVORITE]: unfavoriteName,
  [FAVSCREEN_NAME_UNFAVORITE]: unfavoriteName
});
