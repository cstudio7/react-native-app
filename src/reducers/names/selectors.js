export const getNames = (state, { route, screen }) => {
  const names = state.names[route.key];
  if (screen === 'Favorites') {
    return names.filter(name => name.isFavorite);
  }
  return names;
};