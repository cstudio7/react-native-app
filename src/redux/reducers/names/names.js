const R = require('ramda');

import Actions from '../../../constants/Actions';

export const initialState = {
  female: [],
  male: []
};

const updateName = (name, gender, state) => {
  const genderData = state[gender];
  const index = R.findIndex(R.propEq('id', name.id))(genderData);

  return {
    ...state,
    [gender]: [...genderData.slice(0, index), name, ...genderData.slice(index + 1)]
  };
};

export default (state = initialState, action) => {
  const { type, gender } = action;

  if (type === Actions.saveNames) {
    return {
      ...state,
      [gender]: R.uniq([...state[gender], ...action.names])
    };
  }

  if (type === Actions.favoriteName) {
    return updateName({ ...action.name, isFavorite: true }, gender, state);
  }

  if (type === Actions.unfavoriteName) {
    return updateName({ ...action.name, isFavorite: false }, gender, state);
  }

  return state;
};
