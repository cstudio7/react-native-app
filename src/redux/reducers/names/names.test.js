import names, { initialState } from './names';
import Actions from '../../../constants/Actions';

it('should return the initial state', () => {
  expect(names(undefined, {})).toMatchSnapshot();
});

it('should handle SAVE_NAMES action', () => {
  expect(
    names(initialState, {
      type: Actions.saveNames,
      gender: 'male',
      names: [{ name: 'artur', isFavorite: true }]
    })
  ).toMatchSnapshot();
});

it('should handle FAVORITE_NAME action', () => {
  const state = {
    female: [],
    male: [{ name: 'artur', isFavorite: false }]
  };

  expect(
    names(state, {
      type: Actions.favoriteName,
      gender: 'male',
      name: { name: 'artur', isFavorite: false }
    })
  ).toMatchSnapshot();
});

it('should handle UNFAVORITE_NAME action', () => {
  const state = {
    female: [],
    male: [{ name: 'artur', isFavorite: true }]
  };

  expect(
    names(state, {
      type: Actions.unfavoriteName,
      gender: 'male',
      name: { name: 'artur', isFavorite: true }
    })
  ).toMatchSnapshot();
});
