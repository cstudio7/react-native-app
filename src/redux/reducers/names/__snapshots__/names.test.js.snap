// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle FAVORITE_NAME action 1`] = `
Object {
  "female": Array [],
  "male": Array [
    Object {
      "isFavorite": true,
      "name": "artur",
    },
  ],
}
`;

exports[`should handle SAVE_NAMES action 1`] = `
Object {
  "female": Array [],
  "male": Array [
    Object {
      "isFavorite": true,
      "name": "artur",
    },
  ],
}
`;

exports[`should handle UNFAVORITE_NAME action 1`] = `
Object {
  "female": Array [],
  "male": Array [
    Object {
      "isFavorite": false,
      "name": "artur",
    },
  ],
}
`;

exports[`should return the initial state 1`] = `
Object {
  "female": Array [],
  "male": Array [],
}
`;
